modules/datastore/src/DataDictionary/AlterTableQuery/Builder.php

```diff
@@ -0,0 +1,218 @@
+   *       ],
+   *       [
+   *         'name' => 'field_b',
+   *         'length' => 8 // optional
+   *       ],
+   *     ]
+   *   ]
+   * ]
+   * @endcode
+   */
+  protected array $indexes = [];
+
+  /**
+   * Create an alter table query factory.
+   */
+  public function __construct(
+    DatabaseConnectionFactoryInterface $database_connection_factory,
+    ConverterInterface $date_format_converter,
+    string $query_class
+  ) {
+    $this->databaseConnectionFactory = $database_connection_factory;
+    $this->dateFormatConverter = $date_format_converter;
+    $this->queryClass = $query_class;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setConnectionTimeout(int $timeout): self {
+    $this->databaseConnectionFactory->setConnectionTimeout($timeout);
+
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setTable(string $table): self {
+    $this->table = $table;
+
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addDataDictionary(RootedJsonData $dictionary): self {
+    $this->addFields($dictionary->{'$.data.fields'});
+    // @todo Uncomment once index support has been added to data-dictionaries.
+    // $this->addIndexes($dictionary->{'$.data.indexes'});
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addFields(array $fields): self {
+    // Validate and set fields.
+    $this->fields = array_map([$this, 'validateField'], $fields);
+
+    return $this;
+  }
+
+  /**
+   * Validate data-dictionary field.
+   *
+   * Validate required properties, and fill empty optional properties.
+   *
+   * @param array $field
+   *   Field to validate.
+   *
+   * @return array
+   *   Validated field.
+   */
+  protected function validateField(array $field): array {
+    // Validate required properties.
+    if (empty($field['name']) || empty($field['type'])) {
+      throw new \UnexpectedValueException('"name" and "type" are required properties for data-dictionary fields.');
+    }
+
+    // Provide default values for optional properties.
+    $field['title'] ??= '';
+    $field['format'] ??= '';
+
+    return $field;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function addIndexes(array $indexes): self {
+    // Validate and set indexes.
+    $this->indexes = array_map([$this, 'validateIndex'], $indexes);
+
+    return $this;
+  }
+
+  /**
+   * Validate data-dictionary index.
+   *
+   * Validate required properties, and fill empty optional properties.
+   *
+   * @param array $index
+   *   Index to validate.
+   *
+   * @return array
+   *   Validated index.
+   */
+  public function validateIndex(array $index): array {
+    // Validate nested properties.
+    foreach ($index['fields'] as $key => $index_field) {
+      // Validate required properties.
+      if (empty($index_field['name'])) {
+        throw new \UnexpectedValueException('"name" is a required property for data-dictionary index fields.');
```

Comment:
I would be fine using a InvalidArgumentException, but I do think we benefit from having a layer of validation in the builder, since we can't predict how this will be used in the future.